File: 34_search_for_a_range/searchRange.py
class Solution(object):
    def searchRange(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype List[int]
        """
        if target not in nums:
            return [-1, -1]
        start = 0
        end = len(nums) - 1
        while True:
            pin = (start + end) / 2
            if nums[pin] == target:
                break

            if nums[pin] < target:
                start = pin + 1
            else:
                end = pin - 1
        
        forward = pin
        backward = pin
        
        while forward < len(nums)  and nums[forward] == target:
            forward += 1
        else:
            forward -= 1
        while backward >= 0 and nums[backward] == target:
            backward -= 1
        else:
            backward += 1
        
        return [backward, forward]




if __name__ == '__main__':
    sol = Solution()
    nums = [5, 7, 7, 7, 7, 8, 8, 8, 10]
    target = 9
    print sol.searchRange(nums, target)

File: 34_search_for_a_range/searchRange.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
    public:
        vector<int> searchRange(vector<int>& nums, int target) {
            int begin = 0, end = nums.size() - 1, res = -1;
            while(begin <= end) {
                int mid = begin + (end - begin) / 2;
                if(nums[mid] == target) {
                    res = mid;
                    break;
                }
                if(nums[mid] > target) end = mid - 1;
                else begin = mid + 1;
            }
            if(res == - 1) res = begin;
            if(nums[res] != target) return vector<int>{-1, -1};
            int tmp = res;
            while(tmp < nums.size() && nums[tmp] == target) tmp++;
            while(res > -1 && nums[res] == target) res--;
            return vector<int>{res + 1, tmp - 1};
        }
};

int main() {
    return 0;
}
